ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+    {
+        static_assert(std::is_same_v<tex_ass_type, SContext::tex_ass_type>, "These types must be same!");
+
+        auto found = m_loaderContext->textureCache.find(_element);
+        if (found == m_loaderContext->textureCache.end())
+            return tex_ass_type(nullptr, nullptr, 0.f);
+
+        return found->second;
+    }
+
+    asset::material_compiler::IR::INode* CMitsubaMaterialCompilerFrontend::compileToIRTree(asset::material_compiler::IR* ir, const CElementBSDF* _bsdf)
+{
+    using namespace asset;
+    using namespace material_compiler;
+
+    auto getFloatOrTexture = [this](const CElementTexture::FloatOrTexture& src, IR::INode::SParameter<float>& dst)
+    {
+        if (src.value.type == SPropertyElementData::INVALID)
+        {
+            dst.source = IR::INode::EPS_TEXTURE;
+            //make sure smart_refctd_ptr assignment ptr wont try to drop() -- .value is union
+            dst.value.constant = 0.f;
+            std::tie(dst.value.texture.image, dst.value.texture.sampler, dst.value.texture.scale) = getTexture(src.texture);
+            if (!dst.value.texture.image) {
+                assert(!dst.value.texture.sampler);
+                dst.source = IR::INode::EPS_CONSTANT;
+                dst.value.constant = 0.f;//0 in case when didnt find texture
+            }
+
+        }
+        else
+        {
+            dst.source = IR::INode::EPS_CONSTANT;
+            dst.value.constant = src.value.fvalue;
+        }
+    };
+    auto getSpectrumOrTexture = [this](const CElementTexture::SpectrumOrTexture& src, IR::INode::SParameter<IR::INode::color_t>& dst)
+    {
+        if (src.value.type == SPropertyElementData::INVALID)
+        {
+            dst.source = IR::INode::EPS_TEXTURE;
+            //make sure smart_refctd_ptr assignment ptr wont try to drop() -- .value is union
+            dst.value.constant = IR::INode::color_t(0.f);
+            std::tie(dst.value.texture.image, dst.value.texture.sampler, dst.value.texture.scale) = getTexture(src.texture);
+            if (!dst.value.texture.image) { //TODO should be using loader override here
+                assert(!dst.value.texture.sampler);
+                dst.source = IR::INode::EPS_CONSTANT;
+                dst.value.constant = IR::INode::color_t(1.f, 0.f, 0.f);//red in case when didnt find texture
+            }
+        }
+        else
+        {
+            dst.source = IR::INode::EPS_CONSTANT;
+            dst.value.constant = src.value.vvalue;
+        }
+    };
+
+    constexpr IR::CMicrofacetSpecularBSDFNode::E_NDF ndfMap[4]{
+        IR::CMicrofacetSpecularBSDFNode::ENDF_BECKMANN,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_GGX,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_PHONG,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY
+    };
+
+    IR::INode* root = ir->allocRootNode<IR::CMaterialNode>();
+    root->children.count = 1u;
```

Comment:
what happens if your BSDF AST is just one node?

i.e. just diffuse?

it looks to me like you will allocate another node, and add that as a child

Reply:
what else would i do?

Reply:
ok, when I read this I didnt understand that root was a special node.

I just dont understand why root couldn't be the diffuse node directly.

Reply:
?

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r514558166